### ISSUE TITLE
KCVSLog$MessagePuller does not shut down when using the TitanInputFormat in Apache Spark

### ISSUE THREAD

**Issue by skavulya** [thinkaurelius/titan#817]
I encountered an error when using the TitanHBaseInputFormat in Titan 0.5+ with Apache Spark 1.0.0, and HBase 0.98. The Spark context would not shut down due to a runaway KCVSLog$MessagePuller thread that maintained a connection to the underlying graph. 

The bug occurs due to the following sequence of events when using the TitanInputFormat with Spark's HadoopRDD: 
val hBaseRDD = sparkContext.newAPIHadoopRDD(hBaseConfig, classOf[TitanHBaseInputFormat], classOf[NullWritable], classOf[FaunusVertex])

Spark's HadoopRDD creates a new TitanInputFormat to get the input partitions. TitanInputFormat.setConf() creates a new TitanHBaseHadoopGraph which is opened but never closed. The KCVSLog$MessagePuller maintains a connection to the graph even after the TitanInputFormat instance is out-of-scope in getPartitions() so the Spark context does not fully shut down.

Spark's HadoopRDD's compute method also calls TitanInputFormat.setConf() for each partition.
A new TitanInputFormat is created, along with the corresponding record reader. Once the computation is complete, the Titan graph is shut down correctly using the record reader's close method.

Since close() in Hadoop input formats is associated with the record reader, and not the initial invocation of the input format to get splits, there is currently no way to shutdown the initial graph created.  I worked around it by extending  TitanHBaseInputFormat and closing the graph after getSplits(), but I am not sure if this is the best approach.

**Comment by skavulya** [comment]
A better approach that might solve the issue is to move opening the graph from the TitanHBaseInputFormat.setConf to the TitanHBaseRecordReader.initialize() to ensure that the graph is always closed in case setConf get called multiple times. This approach might also resolve issue  #786 (NullPointerException in TitanHadoop 0.5.1) which I have run across because the reader will now have access to the latest copy of the graph.

**Comment by dalaro** [comment]
Thanks for the report.  I think this is fixed by 5dd656e050cf1f31580a4425e51b3da272d9bfcb, which essentially just what you suggested (for Cassandra as well as HBase).  If you're in a position to try the patch and determine whether it solves the problem in your Spark RDD setup, I'd appreciate the feedback.

**Comment by dalaro** [comment]
I cheated in my first commit by instantiating graphs in `InputFormat.createRecordReader`.  I'm not sure the contract on that method makes this safe.  I moved instantiation into `RecordReader.initialize` with b4eee85165c5c1887d7597a628d5470500ffef8d.

If you're in a position to test, I now suggest applying both 5dd656e050cf1f31580a4425e51b3da272d9bfcb and b4eee85165c5c1887d7597a628d5470500ffef8d.

**Comment by skavulya** [comment]
Thanks Dan! I'll try it out and see if it fixes the issue.

**Comment by skavulya** [comment]
I tested it on Titan 0.5.2. The fix works. Thanks!

**Comment by dalaro** [comment]
Thanks for following up.
